### PR TITLE
Use "volk" instead of statically linked Vulkan loader.

### LINKED ISSUE #51444: Vulkan warning after installing vulkan-sdk probably due outdated Vulkan Loader
### Godot version

4.0.dev.custom_build. f2efa6f4f

### System information

Ubuntu 21.04 - Nvidia GTX 970, Gnome shell 3.38 X11

### Issue description

After installing the latest `vulkan-sdk`, I got this new warnings 
```
ARNING: GENERAL - Message Id Number: 0 | Message Id Name: Loader Message
	loaderAddLayerProperties: /usr/share/vulkan/explicit_layer.d/VkLayer_device_simulation.json invalid layer manifest file version 1.2.0.  May cause errors.
	Objects - 1
		Object[0] - VK_OBJECT_TYPE_INSTANCE, Handle 182720000
     at: _debug_messenger_callback (drivers/vulkan/vulkan_context.cpp:154)
WARNING: GENERAL - Message Id Number: 0 | Message Id Name: Loader Message
	loaderAddLayerProperties: /usr/share/vulkan/explicit_layer.d/VkLayer_khronos_synchronization2.json invalid layer manifest file version 1.2.0.  May cause errors.
	Objects - 1
		Object[0] - VK_OBJECT_TYPE_INSTANCE, Handle 182720000
     at: _debug_messenger_callback (drivers/vulkan/vulkan_context.cpp:154)
```

looks that this warnings should be fixed after updating Vulkan Loader
https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/2729
https://github.com/KhronosGroup/Vulkan-Loader/pull/567

### Steps to reproduce

Run Godot with validation layers installed

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by bruvzg**
Tested so far:
- Building and running on macOS (with and without `use_static_mvk`)
- Building for Windows (volk build only, using MinGW), and running it on wine and Windows 10
- Building for iOS (with static MoltenVK only), and running on simulator.
- Building for Linux (tested volk build only) and running (using swiftshader).
- Building for Android (both with volk and NDK loader), did not tested it on device.

Fixes #51444

### CODE REVIEW

File: thirdparty/README.md [hunk 1]
```diff
@@ -665,6 +665,17 @@ They can be reapplied using the patches included in the `vhacd`
...
+
+- Upstream: https://github.com/zeux/volk
+- Version: 1.2.170.0 (b18f3192886981f28d17f8f41c26c5a53ac3b977, 2021)
+- License: MIT
+
+Files extracted from upstream source:
+
+- `volk.h`, `volk.c`
+- `LICENSE.md`
+
```
Comment (akien-mga) [inline thirdparty/README.md:678]:
```suggestion
- `volk.h`, `volk.c`
- `LICENSE.md`


```
I keep two lines between sections (debatable style choice but it's good to stay consistent :P).

File: thirdparty/README.md [hunk 2]
```diff
@@ -665,6 +665,17 @@ They can be reapplied using the patches included in the `vhacd`
...
 ## vulkan
 
 - Upstream: https://github.com/KhronosGroup/Vulkan-Loader
```
Comment (akien-mga) [inline thirdparty/README.md:681]: We should probably change upstream and track https://github.com/KhronosGroup/Vulkan-Headers directly.

File: drivers/vulkan/SCsub [hunk 3]
```diff
@@ -3,123 +3,91 @@
...
 if env["platform"] == "android":
     # Use NDK Vulkan headers
-    thirdparty_dir = env["ANDROID_NDK_ROOT"] + "/sources/third_party/vulkan/src"
+    thirdparty_dir = "#thirdparty/vulkan"
     thirdparty_includes = [
+        thirdparty_volk_dir,
         thirdparty_dir,
         thirdparty_dir + "/include",
         thirdparty_dir + "/layers",
         thirdparty_dir + "/layers/generated",
```
Comment (akien-mga) [inline drivers/vulkan/SCsub:16]: We don't ship those files, but they were in the NDK.

File: platform/iphone/detect.py [hunk 4]
```diff
@@ -32,7 +32,7 @@ def get_opts():
             "use_static_mvk",
             "Link MoltenVK statically as Level-0 driver (better portability) or use Vulkan ICD loader (enables"
             " validation layers)",
-            False,
+            True,
```
Comment (akien-mga) [inline platform/iphone/detect.py:35]: `use_static_mvk` isn't used anywhere in this script.

File: thirdparty/README.md [hunk 5]
```diff
@@ -665,6 +665,17 @@ They can be reapplied using the patches included in the `vhacd`
 folder.
 
 
+## volk
+
+- Upstream: https://github.com/zeux/volk
+- Version: 1.2.170.0 (b18f3192886981f28d17f8f41c26c5a53ac3b977, 2021)
```
Comment (akien-mga) [inline thirdparty/README.md:671]: I'd think with https://github.com/zeux/volk/commit/d75c007f375f35612dba3de512ac73f10bf9aa0e, `1.2.182` is the latest SDK release in the Vulkan stack: https://github.com/KhronosGroup/Vulkan-Loader/releases/tag/sdk-1.2.182.0

Comment (bruvzg) [reply thirdparty/README.md]: Updated `headers` and `volk` and `vk_enum_string_helper` to the 1.2.182.

File: thirdparty/README.md (hunk 1 above)
Comment (bruvzg) [reply thirdparty/README.md:678]: Fixed

File: thirdparty/README.md (hunk 2 above)
Comment (bruvzg) [reply thirdparty/README.md]: Fixed

**Comment by bruvzg** [comment]
I have replaced `use_static_mvk` flag with `use_volk` flag, which is disabled by default on `Android` and `iOS`, and enabled for the rest of platforms.

On `iOS` and `macOS` disabling it will use static MoltenVK, on other platforms use system loader (NDK loader and headers in case of Android).

File: SConstruct [hunk 6]
```diff
@@ -174,10 +174,10 @@ opts.Add(BoolVariable("builtin_pcre2_with_jit", "Use JIT compiler for the built-
 opts.Add(BoolVariable("builtin_recast", "Use the built-in Recast library", True))
 opts.Add(BoolVariable("builtin_rvo2", "Use the built-in RVO2 library", True))
 opts.Add(BoolVariable("builtin_squish", "Use the built-in squish library", True))
-opts.Add(BoolVariable("builtin_vulkan", "Use the built-in Vulkan loader library and headers", True))
 opts.Add(BoolVariable("builtin_xatlas", "Use the built-in xatlas library", True))
 opts.Add(BoolVariable("builtin_zlib", "Use the built-in zlib library", True))
 opts.Add(BoolVariable("builtin_zstd", "Use the built-in Zstd library", True))
+opts.Add(BoolVariable("use_volk", "Use the volk library", True))
```
[the remark this answers is not in the document]
Comment (bruvzg) [reply SConstruct]: Moved

File: drivers/vulkan/SCsub [hunk 7]
```diff
@@ -2,124 +2,72 @@
...
 
     env_thirdparty.add_source_files(thirdparty_obj, vma_sources)
 
-
 env.drivers_sources += thirdparty_obj
 
-
 # Godot source files
-
 driver_obj = []
```
Comment (akien-mga) [inline drivers/vulkan/SCsub:71]: Not very important but those lines are the style I tried to use consistently in all SCsubs.

File: platform/android/detect.py [hunk 8]
```diff
@@ -52,9 +52,7 @@ def get_android_ndk_root():
 
 
 def get_flags():
-    return [
-        ("tools", False),
-    ]
+    return [("tools", False), ("use_volk", False)]
```
Comment (akien-mga) [inline platform/android/detect.py:55]:
```suggestion
    return [
        ("tools", False),
        ("use_volk", False),
    ]
```

File: platform/iphone/detect.py [hunk 9]
```diff
@@ -28,22 +28,14 @@ def get_opts():
...
         BoolVariable("ios_exceptions", "Enable exceptions", False),
         ("ios_triple", "Triple for ios toolchain", ""),
     ]
 
 
 def get_flags():
-    return [
-        ("tools", False),
-    ]
+    return [("tools", False), ("use_volk", False)]
```
Comment (akien-mga) [inline platform/iphone/detect.py:38]:
```suggestion
    return [
        ("tools", False),
        ("use_volk", False),
    ]
```

File: platform/iphone/display_server_iphone.h [hunk 10]
```diff
@@ -41,7 +41,11 @@
 #include "vulkan_context_iphone.h"
 
 #import <QuartzCore/CAMetalLayer.h>
-#include <vulkan/vulkan_metal.h>
+#ifdef USE_VOLK
+#include <volk.h>
+#else
+#include <vulkan/vulkan.h>
```
Comment (akien-mga) [inline platform/iphone/display_server_iphone.h:47]: Is it intended that this is now `vulkan/vulkan.h` and no longer `vulkan_metal.h`? (same for `vulkan_ios.h`)

File: platform/iphone/detect.py [hunk 11]
```diff
@@ -162,50 +152,15 @@ def configure(env):
...
-        env.Append(LINKFLAGS=["-arch", "armv7", "-Wl,-dead_strip"])
-    if env["arch"] == "arm64":
-        env.Append(LINKFLAGS=["-arch", "arm64", "-Wl,-dead_strip"])
-
-    env.Append(
-        LINKFLAGS=[
-            "-isysroot",
-            "$IPHONESDK",
-        ]
-    )
```
Comment (akien-mga) [inline platform/iphone/detect.py:192]: Nice cleanup if we don't need those anymore, though might be worth doing as a separate commit if it's not related to Volk specifically. Might also be something that can be backported to `3.x`?

File: platform/osx/detect.py [hunk 12]
```diff
@@ -188,12 +182,10 @@ def configure(env):
...
-    if env["use_static_mvk"]:
-        env.Append(LINKFLAGS=["-L$VULKAN_SDK_PATH/MoltenVK/MoltenVK.xcframework/macos-arm64_x86_64/", "-lMoltenVK"])
-        env["builtin_vulkan"] = False
-    elif not env["builtin_vulkan"]:
-        env.Append(LIBS=["vulkan"])
+    if env["vulkan"]:
+        env.Append(CPPDEFINES=["VULKAN_ENABLED"])
+        env.Append(LINKFLAGS=["-framework", "Metal", "-framework", "QuartzCore", "-framework", "IOSurface"])
+        if not env["use_volk"]:
+            env.Append(LINKFLAGS=["-L$VULKAN_SDK_PATH/MoltenVK/MoltenVK.xcframework/macos-arm64_x86_64/", "-lMoltenVK"])
```
Comment (akien-mga) [inline platform/osx/detect.py:189]: So the new default for macOS is to use volk, and if disabled it's to use static MoltenVK IIUC, with no option anymore for dynamic MoltenVK. That works for me :+1: 

How does volk handle MoltenVK? Does it find it from the Vulkan SDK if installed, and from our bundled `.xcframework` otherwise?

File: thirdparty/README.md [hunk 13]
```diff
@@ -660,23 +660,30 @@ They can be reapplied using the patches included in the `vhacd`
 folder.
 
 
+## volk
+
+- Upstream: https://github.com/zeux/volk
+- Version: master (d75c007f375f35612dba3de512ac73f10bf9aa0e, 2021)
```
Comment (akien-mga) [inline thirdparty/README.md:666]:
```suggestion
- Version: git (d75c007f375f35612dba3de512ac73f10bf9aa0e, 2021)
```

File: thirdparty/README.md [hunk 14]
```diff
@@ -660,23 +660,30 @@ They can be reapplied using the patches included in the `vhacd`
...
+- License: MIT
+
+Files extracted from upstream source:
```
Comment (akien-mga) [inline thirdparty/README.md:669]:
```suggestion
The volk commit should match the version of the Vulkan headers defined below.

Files extracted from upstream source:
```

File: thirdparty/vulkan/vk_enum_string_helper.h [hunk 15]
```diff
@@ -4,10 +4,10 @@
```
Comment (akien-mga) [inline thirdparty/vulkan/vk_enum_string_helper.h:4]: You made this file executable by mistake.

**Review by akien-mga — CHANGES_REQUESTED** [review]

File: platform/android/detect.py (hunk 8 above)
Comment (bruvzg) [reply platform/android/detect.py:55]: I was doing it, but it's rejected by static checks CI for some reason.

File: platform/iphone/display_server_iphone.h (hunk 10 above)
Comment (bruvzg) [reply platform/iphone/display_server_iphone.h:47]: Yes. It's auto included in the `vulkan.h` if `VK_USE_PLATFORM_IOS_MVK` is defined, which was missing for iOS and Android and required for volk to build anyway.

File: platform/iphone/detect.py (hunk 11 above)
Comment (bruvzg) [reply platform/iphone/detect.py:192]: It can be back ported. It was never needed.

File: drivers/vulkan/SCsub [hunk 16]
```diff
@@ -2,124 +2,72 @@
...
-    vma_sources = [thirdparty_dir + "/android/vk_mem_alloc.cpp"]
-    env_thirdparty.add_source_files(thirdparty_obj, vma_sources)
 
-elif env["platform"] == "iphone":
-    # Use bundled Vulkan headers
-    thirdparty_dir = "#thirdparty/vulkan"
-    env.Prepend(CPPPATH=[thirdparty_dir, thirdparty_dir + "/include", thirdparty_dir + "/loader"])
+    if env["use_volk"]:
+        env_thirdparty.AppendUnique(CPPDEFINES=["VMA_STATIC_VULKAN_FUNCTIONS=1"])
+        vma_sources = [thirdparty_dir + "/vk_mem_alloc.cpp", thirdparty_volk_dir + "/volk.c"]
```
[the remark this answers is not in the document]
Comment (akien-mga) [reply drivers/vulkan/SCsub:37]: Based on https://github.com/zeux/volk#basic-usage:

> If some files in your application include `vulkan/vulkan.h` and don't include `volk.h`, this can result in symbol conflicts; consider defining `VK_NO_PROTOTYPES` when compiling code that uses Vulkan to make sure this doesn't happen.

and:

> https://github.com/zeux/volk/blob/master/volk.h#L13-L15

Doesn't that mean that it should work provided that we include `volk.h` **before** including `vulkan/vulkan.h`? Or if we define `VK_NO_PROTOTYPES`?

File: platform/osx/detect.py (hunk 12 above)
Comment (bruvzg) [reply platform/osx/detect.py:189]: volk will load try loading `libvulkan.1.dylib` (which is installed as part of SDK), if it's not found it will try `libMoltenVK.dylib`.

xcframework if only for static linking only.

File: drivers/vulkan/SCsub (hunk 16 above)
[the remark this answers is not in the document]
Comment (akien-mga) [reply drivers/vulkan/SCsub:37]: But also, if it works, we might be able to get rid of the `android/vk_mem_alloc.cpp` now and just use the vendored one? It seems the custom one was just to get rid of some warnings, there might be a better way to do it? https://github.com/godotengine/godot/pull/37729#discussion_r406477853

[the remark this answers is not in the document]
Comment (akien-mga) [reply drivers/vulkan/SCsub:37]: > Doesn't that mean that it should work provided that we include `volk.h` before including `vulkan/vulkan.h`? Or if we define `VK_NO_PROTOTYPES`?

BTW I think that means you don't need to patch thirdparty code to use volk, you just need to define `VK_NO_PROTOTYPES` in their build env I think.

[the remark this answers is not in the document]
Comment (bruvzg) [reply drivers/vulkan/SCsub:37]: > This repository contains the Vulkan loader that is used for Linux, Windows, MacOS, and iOS. There is also a separate loader, maintained by Google, which is used on Android.

Seems like Android uses its custom loader, but we probably do not need to use NDK VMA. It's building fine with built-in one, and was added only to suppress some warnings, there are none now. I'll replace it with the built-on one and keep only custom loader headers. That's should fix #51524 build issues as well.

File: platform/android/detect.py (hunk 8 above)
Comment (akien-mga) [reply platform/android/detect.py:55]: It should work if you use a comma at the end.

File: drivers/vulkan/SCsub (hunk 16 above)
[the remark this answers is not in the document]
Comment (bruvzg) [reply drivers/vulkan/SCsub]: > `VK_NO_PROTOTYPES` in their build env I think. Or we have to make sure to include `volk.h` before including them.

Neither of this works, it's failing with "error: redefinition of ... as different kind of symbol" errors.

File: drivers/vulkan/SCsub (hunk 7 above)
Comment (bruvzg) [reply drivers/vulkan/SCsub]: Fixed

File: drivers/vulkan/SCsub (hunk 16 above)
[the remark this answers is not in the document]
Comment (akien-mga) [reply drivers/vulkan/SCsub]: I guess that means that there's still a file which includes `vulkan/vulkan.h` before `volk.h` or before a definition of `VK_NO_PROTOTYPES`. But it's not a big deal, we can have another look at this later, or see with upstream if indeed what they described in the README no longer works with the current SDK version.

File: SConstruct [hunk 17]
```diff
@@ -131,6 +131,7 @@ opts.Add(BoolVariable("xaudio2", "Enable the XAudio2 audio driver", False))
 opts.Add(BoolVariable("vulkan", "Enable the vulkan video driver", True))
 opts.Add("custom_modules", "A list of comma-separated directory paths containing custom modules to build.", "")
 opts.Add(BoolVariable("custom_modules_recursive", "Detect custom modules recursively for each specified path.", True))
+opts.Add(BoolVariable("use_volk", "Use the volk library", True))
```
Comment (akien-mga) [inline SConstruct:134]:
```suggestion
opts.Add(BoolVariable("use_volk", "Use the volk library to load Vulkan drivers", True))
```

Comment (akien-mga) [reply SConstruct:134]: Loading a loading is weird, but that's what it does :P

Comment (bruvzg) [reply SConstruct]: Done

**Comment by akien-mga** [comment]
Small refactoring proposal for the `SCsub`, should be equivalent but it flows better IMO:
```python
#!/usr/bin/env python

Import("env")

if env["use_volk"]:
    env.AppendUnique(CPPDEFINES=["USE_VOLK"])

thirdparty_obj = []
thirdparty_volk_dir = "#thirdparty/volk"
thirdparty_dir = "#thirdparty/vulkan"

if env["platform"] == "android" and not env["use_volk"]:
    # Use NDK Vulkan headers
    ndk_vulkan_dir = env["ANDROID_NDK_ROOT"] + "/sources/third_party/vulkan/src"
    thirdparty_includes = [
        ndk_vulkan_dir,
        ndk_vulkan_dir + "/include",
        ndk_vulkan_dir + "/layers",
        ndk_vulkan_dir + "/layers/generated",
    ]
    env.Prepend(CPPPATH=thirdparty_includes)
else:
    # Use bundled Vulkan headers
    env.Prepend(CPPPATH=[thirdparty_dir, thirdparty_dir + "/include"])
    if env["use_volk"]:
        env.Prepend(CPPPATH=[thirdparty_volk_dir])

if env["platform"] == "android":
    env.AppendUnique(CPPDEFINES=["VK_USE_PLATFORM_ANDROID_KHR"])
elif env["platform"] == "iphone":
    env.AppendUnique(CPPDEFINES=["VK_USE_PLATFORM_IOS_MVK"])
elif env["platform"] == "linuxbsd":
    env.AppendUnique(CPPDEFINES=["VK_USE_PLATFORM_XLIB_KHR"])
elif env["platform"] == "osx":
    env.AppendUnique(CPPDEFINES=["VK_USE_PLATFORM_MACOS_MVK"])
elif env["platform"] == "windows":
    env.AppendUnique(CPPDEFINES=["VK_USE_PLATFORM_WIN32_KHR"])

# Build Vulkan memory allocator and volk
env_thirdparty_vma = env.Clone()
env_thirdparty_vma.disable_warnings()
thirdparty_sources_vma = [thirdparty_dir + "/vk_mem_alloc.cpp"]

if env["use_volk"]:
    env_thirdparty_vma.AppendUnique(CPPDEFINES=["VMA_STATIC_VULKAN_FUNCTIONS=1"])
    env_thirdparty_volk = env.Clone()
    env_thirdparty_volk.disable_warnings()

    thirdparty_sources_volk = [thirdparty_volk_dir + "/volk.c"]
    env_thirdparty_volk.add_source_files(thirdparty_obj, thirdparty_sources_volk)

env_thirdparty_vma.add_source_files(thirdparty_obj, thirdparty_sources_vma)


env.drivers_sources += thirdparty_obj


# Godot source files

driver_obj = []

env.add_source_files(driver_obj, "*.cpp")
env.drivers_sources += driver_obj

# Needed to force rebuilding the driver files when the thirdparty code is updated.
env.Depends(driver_obj, thirdparty_obj)
```

**Comment by akien-mga** [comment]
Last thing and it's good to go:
```diff
diff --git a/COPYRIGHT.txt b/COPYRIGHT.txt
index b2a930a4bb..5bd67960da 100644
--- a/COPYRIGHT.txt
+++ b/COPYRIGHT.txt
@@ -438,12 +438,17 @@ Copyright: 2011, Khaled Mamou
   2003-2009, Erwin Coumans
 License: BSD-3-clause
 
+Files: ./thirdparty/volk/
+Comment: volk
+Copyright: 2018-2019, Arseny Kapoulkine
+License: Expat
+
 Files: ./thirdparty/vulkan/
-Comment: Vulkan Ecosystem Components (Vulkan ICD loader and headers)
-Copyright: 2014-2020, The Khronos Group Inc.
-  2014-2020, Valve Corporation
-  2014-2020, LunarG, Inc.
-  2015-2020, Google Inc.
+Comment: Vulkan Headers
+Copyright: 2014-2021, The Khronos Group Inc.
+  2014-2021, Valve Corporation
+  2014-2021, LunarG, Inc.
+  2015-2021, Google Inc.
 License: Apache-2.0
 
 Files: ./thirdparty/vulkan/vk_mem_alloc.h

```

**Review by akien-mga — APPROVED** [review]
Looks great! Thanks a lot.

**Comment by akien-mga** [comment]
Tested on Linux with both `use_volk=yes` (default) and `use_volk=no` (using system Vulkan Loader, needs to be installed) and both work fine :+1:

**Comment by Calinou** [comment]
Is it possible to implement https://github.com/zeux/volk#optimizing-device-calls in Godot?

**Comment by bruvzg** [comment]
> Is it possible to implement https://github.com/zeux/volk#optimizing-device-calls in Godot?

Probably, I do not see any reason why it won't work, but I have not tested it.

**Comment by zeux** [comment]
FWIW Roblox ships with volk on Android as well (valuable for reduced overhead and extension support), so that should work the same as desktop platforms.

**Comment by akien-mga** [comment]
> FWIW Roblox ships with volk on Android as well (valuable for reduced overhead and extension support), so that should work the same as desktop platforms.

Thanks for the input! We'll do that, that simplifies things and avoids limiting us to whichever version of the headers our current NDK provides.
